### PR TITLE
feat: token management for workers

### PR DESCRIPTION
**Pull request by 3Hren**
@noxiouz PTAL.

TBD: how to notify users about errors when the manager fails to refresh tickets?

- [x] More tests.

# Examples

The next example shows an application, which returns the most fresh TVM ticket.

```go
package main

import (
	"fmt"
	"golang.org/x/net/context"
	"time"

	_ "github.com/cocaine/cocaine-framework-go/auth/tvm"
	"github.com/cocaine/cocaine-framework-go/cocaine12"
)

func main() {
	w, err := cocaine12.NewWorker()
	if err != nil {
		panic(err)
	}

	w.SetTerminationHandler(func(ctx context.Context) {
		time.Sleep(60 * time.Millisecond)
	})

	w.EnableStackSignal(false)
	w.On("token", func(ctx context.Context, req cocaine12.Request, resp cocaine12.Response) {
		defer resp.Close()
		token := w.Token()

		resp.Write([]byte(token.Body()))
	})

	if err = w.Run(nil); err != nil {
		fmt.Printf("%v", err)
	}
}
```

### CODE REVIEW

File: cocaine12/token.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+		select {
+		case <-t.ticker.C:
+			t.mu.Lock()
+			body := t.ticket.Body()
+			t.mu.Unlock()
+
+			ctx := context.Background()
+			ch, err := t.tvm.Call(ctx, "refresh_ticket", t.appName, body)
+			if err != nil {
+				fmt.Printf("failed to update ticket %v\n", err)
```
Comment (3Hren) [inline cocaine12/token.go]: Unsure about it's the best way to notify about errors.

File: cocaine12/token.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+			var ticketResult string
+			if err := answer.ExtractTuple(&ticketResult); err != nil {
+				fmt.Printf("failed to extract ticket %v\n", err)
+				continue
+			}
+
+			t.mu.Lock()
+			t.ticket = &Token{tvmTokenType, ticketResult}
+			t.mu.Unlock()
+		case <-t.stopped:
```
Comment (noxiouz) [inline cocaine12/token.go]: I would use cancellation via Context. Also context.WithTimeout could be passed to Call.

File: cocaine12/token.go [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+			t.ticker.Stop()
+			return
+		}
+	}
+}
+
+func newTokenManager(appName string, token *Token) (TokenManager, error) {
```
Comment (noxiouz) [inline cocaine12/token.go]: Public to use w/o Worker related stuff.

File: cocaine12/workerng.go [hunk 4]
```diff
@@ -125,18 +127,25 @@ func NewWorkerNG() (*WorkerNG, error) {
 			unixSocketEndpoint, err)
 	}
 
+	tokenManager, err := newTokenManager(GetDefaults().ApplicationName(), GetDefaults().Token())
+	if err != nil {
```
Comment (noxiouz) [inline cocaine12/workerng.go]: sock.Close()

File: cocaine12/token.go [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+// Token carries information about an authorization token.
+type Token struct {
+	ty   string
+	body string
+}
+
+// Type returns token's type.
+// Depending on the context it can be empty, OAUTH, TVM etc.
+func (t *Token) Type() string {
```
Comment (noxiouz) [inline cocaine12/token.go]:
```go
func (t Token) Type() string
```
strings are immutable so Token is fast to be copied and no need to return a pointer to Token in tokenManager.Token()

Comment (noxiouz) [inline cocaine12/token.go]: same for Body()

**Comment by noxiouz** [comment]
LGTM

**Pull request merged**